File: Algorithm/Python/052. N-Queens II.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-04-28 23:10:30
# @Last modified by:   WuLC
# @Last Modified time: 2016-04-28 23:10:48
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        matrix  = [['.' for i in xrange(n)] for j in xrange(n)]
        result = [0]
        self.helper(matrix,0,result)
        return result[0]
    
    def helper(self,matrix,row,result):
        n = len(matrix)
        if row == n:
            result[0] += 1
            return 
        for i in xrange(n):
            if self.is_valid(matrix,row,i):
                matrix[row][i] = 'Q'
                self.helper(matrix,row+1,result)
                matrix[row][i] = '.'
        
                
    def is_valid(self,matrix,i,j):
        n = len(matrix)
    
        for k in xrange(n):
            if matrix[k][j] == 'Q':
                return False
        
        ti,tj=i,j
        # left
        while ti>=0 and tj>=0:
            if matrix[ti][tj] == 'Q':
                return False
            ti -= 1
            tj -= 1
        # right
        while 0<=i<n and 0<=j<n:
            if matrix[i][j]=='Q':
                return False
            i -= 1
            j += 1
        
        return True

File: Algorithm/Java/052. N-Queens II.java
/**
* Author: WuLC
* Date:   2016-12-27 23:22:58
* Last modified by:   WuLC
* Last Modified time: 2016-12-28 00:18:33
* Email: liangchaowu5@gmail.com
*/


// method 1, dfs
// with bit manipulation instead of boolean array to judge collision
public class Solution 
{
    private int result, column, leftBias, rightBias;
    public int totalNQueens(int n) 
    {
        result = column = leftBias = rightBias = 0;
        dfs(0, n);
        return result;
    }
    
    public void dfs(int row, int n)
    {
        if (row == n)
        {
            result += 1;
            return;
        }
        for(int col = 0; col < n; col++)
        {
            if ((column&(1<<col)) == 0 && (leftBias&(1<<(row+col))) == 0 && (rightBias&(1<<(row-col+n-1))) == 0)
            {
                column ^= (1<<col); leftBias ^= (1<<(row+col)); rightBias ^= (1<<(row-col+n-1));
                dfs(row+1,n);
                column ^= (1<<col); leftBias ^= (1<<(row+col)); rightBias ^= (1<<(row-col+n-1));
            }
        }
        
    }
}


// method 2, dfs
// base on method 1, find the valid positions for each row instead of traverse and judge one by one
public class Solution 
{
    private int result, column, leftBias, rightBias;
    public int totalNQueens(int n) 
    {
        result = column = leftBias = rightBias = 0;
        dfs(0, n);
        return result;
    }
    
    public void dfs(int row, int n)
    {
        if (row == n)
        {
            result += 1;
            return;
        }
        int validPosition =  (~(column | (leftBias >> row) | (rightBias >> (n-row-1)))) & ((1<<n)-1);
        while (validPosition != 0)
        {   
            int col = validPosition&(-validPosition); // low bit 
            validPosition ^= col;
            column ^= col; leftBias ^= (col << row) ; rightBias ^= (col << (n - row-1));
            dfs(row+1,n);
            column ^= col; leftBias ^= (col << row) ; rightBias ^= (col << (n - row-1));
        }
    }
}


